src/transform/src/reduction_pushdown.rs

```diff
@@ -155,6 +157,149 @@ impl ReductionPushdown {
+
+        let mut builder = DerivedBuilder::new(ctx.features);
+        builder.require(Arity);
+        builder.require(RelationType);
+        builder.require(Equivalences);
+        let derived = builder.visit(relation);
+        let derived = derived.as_view();
+
+        let mut todo = vec![(relation, derived)];
+        while let Some((expr, view)) = todo.pop() {
+            if let Some(replacement) = Self::assess_disjointness(expr, view) {
+                *expr = replacement;
+            } else {
+                todo.extend(expr.children_mut().rev().zip(view.children_rev()));
+            }
+        }
+    }
+
+    fn assess_disjointness(expr: &MirRelationExpr, view: DerivedView) -> Option<MirRelationExpr> {
+        use crate::analysis::equivalences::Equivalences;
+        use crate::analysis::{Arity, RelationType};
+
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            expected_group_size,
+            ..
+        } = expr
+        {
+            if let MirRelationExpr::Union { .. } = &**input {
+                // Form the union input equivalences restricted to the `group_key` expressions.
+                let arity = *view.last_child().value::<Arity>().expect("Arity required");
+                let child_equivs = view
+                    .last_child()
+                    .children_rev()
+                    .map(|view| {
+                        let mut equivs = view
+                            .value::<Equivalences>()
+                            .expect("Equivalences required")
+                            .clone();
+                        if let Some(equivs) = &mut equivs {
+                            let types = view
+                                .value::<RelationType>()
+                                .expect("RelationType required")
+                                .as_ref()
+                                .unwrap()
+                                .clone();
+                            for (index, key_expr) in group_key.iter().enumerate() {
+                                equivs.classes.push(vec![
+                                    key_expr.clone(),
+                                    MirScalarExpr::column(arity + index),
+                                ]);
+                            }
+                            equivs.minimize(&Some(types.clone()));
+                            equivs.project(arity..arity + group_key.len());
+                            equivs.minimize(&Some(types.clone()));
+                        }
+                        equivs
+                    })
+                    .collect::<Vec<_>>();
+
+                // Assess the pairwise overlap of the equivalences.
+                let mut potential_overlap = Vec::new();
+                for i in 0..child_equivs.len() {
+                    if let Some(equiv_i) = &child_equivs[i] {
+                        for j in 0..i {
```

Comment:
The quadraticness here seems a bit scary. I think we've seen Unions with many inputs.